daml/Marketplace/Distribution/Auction/Service.daml

```diff
@@ -0,0 +1,201 @@
+        assertMsg ("Unexpected lead bidder - " <> show bid.customer) (bid.customer == auction.leadBidder)
+
+        let
+          allocationPercentage = roundBankers 2 $ bid.details.quantity / auction.asset.quantity
+          leadBidder = bid.customer
+        create LeadBidder with ..
+
+    controller customer can
+      nonconsuming RequestCreateAuction : ContractId CreateAuctionRequest
+        with
+          leadBidder : Party
+          asset : Asset
+          quotedAssetId : Id
+          assetDepositCid : ContractId AssetDeposit
+          allocationAccount : Account
+          receivableAccount : Account
+          floorPrice : Decimal
+        do
+          create CreateAuctionRequest with ..
+
+      nonconsuming CancelAuctionRequest : ()
+        with
+          createAuctionRequestCid : ContractId CreateAuctionRequest
+        do
+          archive createAuctionRequestCid
+
+    controller provider can
+      nonconsuming CreateAuction : ContractId Auction.T
+        with
+          createAuctionRequestCid : ContractId CreateAuctionRequest
+        do
+          CreateAuctionRequest{..} <- fetchAndArchive createAuctionRequestCid
+
+          -- Deposit the asset into the allocation account so they cannot be transferred/withdrawn without the signatore of the Auctioneer
+          depositCid <- exerciseByKey @AllocationAccountRule allocationAccount.id Allocation.Deposit with depositCid = assetDepositCid; ..
+          create Auction with status = Auction.Open; ..
+
+      nonconsuming RejectAuction : ()
+        with
+          createAuctionRequestCid : ContractId CreateAuctionRequest
+        do
+          archive createAuctionRequestCid
+
+      nonconsuming ProcessAuction : (ContractId Auction.T, [ContractId Settlement.SettlementInstruction])
+        with
+          auctionCid : ContractId Auction.T
+          bidCids : [ContractId Bid.T]
+        do
+          auction@Auction{..} <- fetch auctionCid
+          assertMsg "Auction already successfully processed" (status `elem` [Auction.Open, Auction.NoValidBids])
+          bids <- forA bidCids fetch
+
+          -- Split bids on those above / below the floor price -> Sort valid bids descendingly -> Execute Auction -> Split auction result on allocated / unallocated (ie, zero allocation)
+          let
+            (validBids, invalidBids) = splitList (\bid -> bid.details.price < floorPrice) bids
+            sortedBids = sortOn (\bid -> Down bid.details) validBids
+            ((remaining, finalPrice), allocations) = mapAccumL dutchAuction (asset.quantity, 0.0) sortedBids
+            (allocatedBids, unallocatedBids) = splitList (\allocation -> allocation.quantity == 0.0) allocations
+
+          -- Return assets of all invalid and unallocated bids. Also update their bid status accordingly
+          forA_ invalidBids $ withdrawAllocation *> updateBidStatus Bid.Invalid
+          forA_ ((.bid) `map` unallocatedBids) $ withdrawAllocation *> updateBidStatus Bid.NoAllocation
```

Comment:
Not a big deal, but I'm not too keen on these side effecting (i.e. returning unit) functions.

Reply:
You know I'm not a fan of adding short-lived types, but in this case I might make an exception and return the invalid and not allocated bids, altogether in a new `AuctionResult`.
I think you may be doing the UI a favour by providing easy access on these, instead of having to query for them.

Reply:
As discussed - @GeorgSchneider is to look at this as he's working on the UI.